packages/gatsby-image/src/index.js

```diff
@@ -160,6 +185,17 @@ class Image extends React.Component {
       IOSupported = true
     }
 
+    // Chrome Canary 75 added native lazy loading support!
+    // https://addyosmani.com/blog/lazy-loading/
+    if (
+      typeof HTMLImageElement !== `undefined` &&
```

Comment:
You could set this immediately after declaring `nativeLazyLoadSupported` (or use the value of this to define that variable).

```JS
// Extracted feature detection to vars to avoid the interleaving logic flipping booleans, especially `isVisible`
// Native lazy-loading support: https://addyosmani.com/blog/lazy-loading/
const hasNativeLazyLoadSupport = (
  typeof HTMLImageElement !== `undefined` &&
  `loading` in HTMLImageElement.prototype
)

const hasIOSupport = (
  typeof window !== `undefined` &&
  window.IntersectionObserver
)

// Renamed `hasNoScript` and `IOSupport` vars to have clearer intents
const isBrowser = typeof window !== `undefined`
const addNoScript = !(props.critical && !props.fadeIn)
const useIOSupport = !props.critical && !seenBefore && hasIOSupport

// Condensed all the prior logic for setting `isVisible`, ordered by override priority
// Probably isn't that readable and would be better off verbose?
const isVisible = (
  props.critical || 
  isBrowser && (hasNativeLazyLoadSupport || !useIOSupport)
)
```

Too terse? I think it's a bit clearer regarding the interplay vs the conditional statements flipping booleans, but extensive short-circuit via logical operators might counter readability benefits, especially for `isVisible`?

Reply:
```JS
// Visible if `critical` or is in a browser with native lazy-load support, or no IntersectionObserver support available. Not visible/rendered during SSR(unless `critical`), or when lazy-loading with IntersectionObserver in the browser.
const isVisible = (
  props.critical || 
  isBrowser && (hasNativeLazyLoadSupport || !useIOSupport)
)
```

Perhaps the comment clears up any confusion/thinking.. or maybe it's just bad code on my part :sweat_smile: